sharedCache broken for rules DSL

The shared cache is no longer available in rules DSL?


[The test rule mentioned here](https://community.openhab.org/t/unable-to-use-private-cache-in-rules-dsl/143552/5)
results in

2024-11-12 11:04:00.271 [ERROR] [ore.automation.module.script.internal.handler.ScriptActionHandler] - Script execution of rule with UID 'x-1' failed: The name 'sharedCache' cannot be resolved to an item or type; line 5, column 27, length 11 in x


There's hints on the forum it worked in 4.1 and broke somewhere between 4.1 and 4.2.

Related: #3187

--- a/bundles/org.openhab.core.model.rule/bnd.bnd
+++ b/bundles/org.openhab.core.model.rule/bnd.bnd
@@ -31,6 +31,7 @@ Import-Package: \
  org.openhab.core.model.persistence.extensions,\
  org.openhab.core.model.script,\
  org.openhab.core.model.script.engine.action,\
+ org.openhab.core.automation.module.script.rulesupport.shared,\
  com.google.common.base;version="14",\
  javax.measure,\
  javax.measure.quantity,\

--- a/bundles/org.openhab.core.model.rule/pom.xml
+++ b/bundles/org.openhab.core.model.rule/pom.xml
@@ -20,6 +20,11 @@
       <artifactId>org.openhab.core.model.script</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>org.openhab.core.bundles</groupId>
+      <artifactId>org.openhab.core.automation.module.script.rulesupport</artifactId>
+      <version>${project.version}</version>
+    </dependency>
   </dependencies>
 
   <build>

--- a/bundles/org.openhab.core.model.rule/src/org/openhab/core/model/rule/jvmmodel/RulesJvmModelInferrer.xtend
+++ b/bundles/org.openhab.core.model.rule/src/org/openhab/core/model/rule/jvmmodel/RulesJvmModelInferrer.xtend
@@ -41,6 +41,7 @@ import org.slf4j.LoggerFactory
 import org.eclipse.xtext.common.types.JvmFormalParameter
 import org.eclipse.emf.common.util.EList
 import org.openhab.core.events.Event
+import org.openhab.core.automation.module.script.rulesupport.shared.ValueCache
 
 /**
  * <p>Infers a JVM model from the source model.</p> 
@@ -130,6 +131,10 @@ class RulesJvmModelInferrer extends ScriptJvmModelInferrer {
             members += ruleModel.rules.map [ rule |
                 rule.toMethod("_" + rule.name, ruleModel.newTypeRef(Void.TYPE)) [
                     static = true
+                    val privateCacheTypeRef = ruleModel.newTypeRef(ValueCache)
+                    parameters += rule.toParameter(VAR_PRIVATE_CACHE, privateCacheTypeRef)
+                    val sharedCacheTypeRef = ruleModel.newTypeRef(ValueCache)
+                    parameters += rule.toParameter(VAR_SHARED_CACHE, sharedCacheTypeRef)
                     if ((containsCommandTrigger(rule)) || (containsStateChangeTrigger(rule)) || (containsStateUpdateTrigger(rule))) {
                         val groupTypeRef = ruleModel.newTypeRef(Item)
                         parameters += rule.toParameter(VAR_TRIGGERING_GROUP, groupTypeRef)

--- a/bundles/org.openhab.core.model.script.runtime/src/org/openhab/core/model/script/runtime/internal/engine/DSLScriptEngine.java
+++ b/bundles/org.openhab.core.model.script.runtime/src/org/openhab/core/model/script/runtime/internal/engine/DSLScriptEngine.java
@@ -169,8 +169,10 @@ public class DSLScriptEngine implements javax.script.ScriptEngine {
 
         Map<String, Object> cachePreset = scriptExtensionAccessor.findPreset("cache",
                 (String) context.getAttribute("oh.engine-identifier", ScriptContext.ENGINE_SCOPE));
-        evalContext.newValue(QualifiedName.create("sharedCache"), cachePreset.get("sharedCache"));
-        evalContext.newValue(QualifiedName.create("privateCache"), cachePreset.get("privateCache"));
+        evalContext.newValue(QualifiedName.create(ScriptJvmModelInferrer.VAR_SHARED_CACHE),
+                cachePreset.get("sharedCache"));
+        evalContext.newValue(QualifiedName.create(ScriptJvmModelInferrer.VAR_PRIVATE_CACHE),
+                cachePreset.get("privateCache"));
         // now add specific implicit vars, where we have to map the right content
         Object value = context.getAttribute(OUTPUT_EVENT);
         if (value instanceof ChannelTriggeredEvent event) {

